substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVFileRecord.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2020, 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020, Red Hat Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.objectfile.pecoff.cv;
+
+import com.oracle.objectfile.debugentry.FileEntry;
+
+import java.io.PrintStream;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Paths;
+import java.security.MessageDigest;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+final class CVFileRecord extends CVSymbolRecord {
+
+    private static final byte CHECKSUM_NONE = 0x00;
+    private static final byte CHECKSUM_MD5 = 0x01;
+    private static final byte CB_VALUE = 0x10;
+
+    private static final int FILE_RECORD_LENGTH = 24;
+
+    private static final int CHECKSUM_LENGTH = 16;
+    private static final byte[] EMPTY_CHECKSUM = new byte[CHECKSUM_LENGTH];
+
+    private static final int FILE_TABLE_INITIAL_SIZE = 200;
+
+    private final CVSymbolSectionImpl.CVStringTable strings;
+
+    private int currentOffset = 0;
+    private Map<FileEntry, Integer> fileEntryToOffsetMap = new LinkedHashMap<>(FILE_TABLE_INITIAL_SIZE);
+
+    CVFileRecord(CVDebugInfo cvDebugInfo, CVSymbolSectionImpl.CVStringTable strings) {
+        super(cvDebugInfo, CVDebugConstants.DEBUG_S_FILECHKSMS);
+        this.strings = strings;
+    }
+
+    /*
+     * Convert a simple path into an absolute path by determining if it's part of Graal, the JDK, or
+     * use code.
+     *
+     * Currently, don't even try; use the SourceCache system
+     */
+    private static String fixPath(FileEntry fileEntry) {
```

Comment:
I'm not sure this is going to be needed? is there a reason why debugging the native image in a project that uses the cached source bundle is going to be unsatisfactory?
Remember once again that the target audience for this debug capability is not developers during developement of an app but developers trying to debug a problem in a deployed native app that has already been tested and shown to work on OpenJDK. So, trying to get the debug info back to refer back to an existing dev project source tree should not really be a concern?

Reply:
I've rewritten the comment on this (and coded it better).  It protects against an NPE when there is no DirEntry in the FileEntry.  I don't remember the circumstances, but it happens in this code.

Reply:
Hmm, I was not expecting that to happen but I suppose it might occur for a class in the default package.
I'm not sure this is the right place to fix it though. I think it would be better to fix FileEntry.getFullName to detect that dirEntry is null and in that case just return getFileName().